File: JavaScript/boj/Silver/1780.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const [N, ...input] = fs.readFileSync(filePath).toString().trim().split('\n');

const paper = [];
const result = [0, 0, 0];

for (let i = 0; i < N; i++) paper[i] = input[i].split(' ').map(Number);

const isSamePaper = (row, col, size) => {
  const start = paper[row][col];

  for (let i = row; i < row + size; i++) {
    for (let j = col; j < col + size; j++) {
      if (start !== paper[i][j]) return false;
    }
  }

  return true;
};

const divide = (row, col, size) => {
  if (!isSamePaper(row, col, size)) {
    const divideSize = size / 3;

    divide(row, col, divideSize);
    divide(row, col + divideSize, divideSize);
    divide(row, col + 2 * divideSize, divideSize);
    divide(row + divideSize, col, divideSize);
    divide(row + divideSize, col + divideSize, divideSize);
    divide(row + divideSize, col + 2 * divideSize, divideSize);
    divide(row + 2 * divideSize, col, divideSize);
    divide(row + 2 * divideSize, col + divideSize, divideSize);
    divide(row + 2 * divideSize, col + 2 * divideSize, divideSize);
  } else result[paper[row][col] + 1] += 1;
};

divide(0, 0, N);
console.log(result.join('\n'));

File: Python/boj/Silver/1780.py
import sys
from collections import Counter


def isSamePaper(row, col, size):
    start = paper[row][col]

    return all(
        paper[i][j] == start
        for i in range(row, row + size)
        for j in range(col, col + size)
    )


def divide(row, col, size):
    if not isSamePaper(row, col, size):
        divideSize = size // 3

        divide(row, col, divideSize)
        divide(row, col + divideSize, divideSize)
        divide(row, col + 2 * divideSize, divideSize)
        divide(row + divideSize, col, divideSize)
        divide(row + divideSize, col + divideSize, divideSize)
        divide(row + divideSize, col + 2 * divideSize, divideSize)
        divide(row + 2 * divideSize, col, divideSize)
        divide(row + 2 * divideSize, col + divideSize, divideSize)
        divide(row + 2 * divideSize, col + 2 * divideSize, divideSize)
    else:
        result[paper[row][col]] += 1


if __name__ == "__main__":
    try:
        sys.stdin = open("input.txt", "r")
    except FileNotFoundError:
        pass

    N = int(input())
    paper = [list(map(int, input().split())) for _ in range(N)]
    result = Counter()

    divide(0, 0, N)

    print(result[-1])
    print(result[0])
    print(result[1])
